Fully unit tested consul-powered storage

lord2800 opened the pull request:
Depends on https://github.com/billimanjaro/billimanjaro/pull/1 which is now ready to be merged.

lord2800 committed:
Fully unit tested consul-powered storage

lord2800 commented:
Important note: `add` and `update` are not atomic operations due to https://github.com/hashicorp/consul/issues/886.

gabelimon commented on src/api/lib/storage/backends/consul.js:
```diff
@@ -0,0 +1,109 @@
+function ConsulBackend(options) {
+	options = options || {};
+	this._prefix = options.prefix || 'billimanjaro';
+	delete options.prefix;
```
Why are you deleting the the array prefix?

lord2800 replied:
So I can send the remaining options, without the prefix (just in case for backwards compatibility), to the consul client.

gabelimon commented on src/api/lib/storage/backends/consul.js:
```diff
@@ -0,0 +1,109 @@
+
+function ConsulBackend(options) {
+	options = options || {};
+	this._prefix = options.prefix || 'billimanjaro';
```
What is `this` in this case?

lord2800 replied:
The instance of the consul storage backend.

gabelimon replied:
This function confused me a bit but I think it's fine. I'll re-read it a bit more.

lord2800 replied:
It's a [constructor](https://developer.mozilla.org/en-US/docs/Glossary/Constructor) in javascript.

gabelimon commented on src/api/lib/storage/backends/consul.js:
```diff
@@ -0,0 +1,109 @@
+function syncEntity(backend, entity, id, object, callback) {
+	if(Object.keys(object).length === 0 ||
+		(Object.keys(object).length === 1 &&
+		Object.keys(object)[0] === 'id')) {
```
Shouldn't you just check if `id` is contained rather than asserting the location in the array?

lord2800 replied:
I suppose, I'm asserting the array length just above that though.

gabelimon replied:
Ahhh, got it.

gabelimon commented on src/api/lib/storage/backends/consul.js:
```diff
@@ -0,0 +1,109 @@
+			callback(err || new Error(response.statusMessage));
+			return;
+		}
+		if(--count === 0 && !hadError) {
```
Couldn't this just be else if without the return?

lord2800 replied:
Good call.

gabelimon commented on src/api/lib/storage/backends/consul.js:
```diff
@@ -0,0 +1,109 @@
+		callback(new Error('Null cannot be saved!'));
+		return;
+	} else if(typeof object !== 'object') {
+		callback(new Error('You must pass an object!'));
```
It would be helpful to dump the object type here.

lord2800 replied:
Agreed.

gabelimon commented on src/api/lib/storage/backends/consul.js:
```diff
@@ -0,0 +1,109 @@
+	} else if(!object.hasOwnProperty('id')) {
+		callback(new Error('Your object must already have an id!'));
+		return;
+	}
```
This whole block could be in a `validate_object` function

lord2800 replied:
Yeah, I can move part of this to a common validate function.

gabelimon replied:
Corrected to block, not function.

lord2800 replied:
The only one of these conditions I can't move is the `id` check--it's specific to `update` vs. `add` (which could assert the lack of an `id`, but simply deletes it if present currently--do you have a preference?).

gabelimon replied:
I'd just leave the ID check as an update-specific requirement.

gabelimon commented:
:+1: all of my comments seem minor.

lord2800 commented:
Wanna give it another once-over @gabelimon? I cleaned up a few minor things and moved the `add`/`update` common validation logic into its own function (with proper error types, even).

gabelimon commented:
:+1:
